src/parquet/column/scanner-test.cc

```diff
@@ -0,0 +1,386 @@
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <cstdint>
+#include <cstdlib>
+#include <list>
+#include <vector>
+#include <limits>
+
+#include <gtest/gtest.h>
+
+#include "parquet/types.h"
+#include "parquet/column/page.h"
+#include "parquet/column/reader.h"
+#include "parquet/column/scanner.h"
+#include "parquet/column/test-util.h"
+
+#include "parquet/util/test-common.h"
+
+namespace parquet_cpp {
+
+using schema::NodePtr;
+
+bool operator==(const Int96 a, const Int96 b) {
+  return a.value[0] == b.value[0] &&
+         a.value[1] == b.value[1] &&
+         a.value[2] == b.value[2];
+}
+
+bool operator==(const ByteArray a, const ByteArray b) {
+  return a.len == b.len && 0 == memcmp(a.ptr, a.ptr, a.len);
+}
+
+static int FLBA_LENGTH = 8;
+bool operator==(const FixedLenByteArray a, const FixedLenByteArray b) {
+  return 0 == memcmp(a.ptr, a.ptr, FLBA_LENGTH);
+}
+
+namespace test {
+
+template <int TYPE>
+class TestFlatScanner : public ::testing::Test {
+ public:
+  void SetUp() {
+    values_.resize(0);
+    std::random_device rd;
+    SEED = rd();
+  }
+
+  void TearDown() {}
+
+  typedef typename type_traits<TYPE>::value_type T;
+
+  void InitScanner(Repetition::type rep,
+      int16_t max_def_level = 0, int16_t max_rep_level = 0) {
+    NodePtr type = schema::PrimitiveNode::Make("c1", rep, static_cast<Type::type>(TYPE));
+    descr_.reset(new ColumnDescriptor(type, max_def_level, max_rep_level));
+
+    std::unique_ptr<PageReader> pager(new test::MockPageReader(pages_));
+    scanner_ = Scanner::Make(ColumnReader::Make(descr_.get(), std::move(pager)));
+  }
+
+ protected:
+  void AddPages(size_t pages, size_t num_values,
+      std::vector<int16_t> def_levels, int16_t max_def_level,
+      std::vector<int16_t> rep_levels, int16_t max_rep_level) {
+    size_t page_size = std::max(def_levels.size(), num_values);
+    std::vector<T> page(num_values);
+    for (size_t i = 0; i < pages; i++) {
+      random_numbers(num_values, SEED, &page[0]);
+      buffers.push_back(std::vector<uint8_t>());
+      pages_.push_back(
+          MakeDataPage<TYPE>(page, def_levels, max_def_level,
+              rep_levels, max_rep_level, &buffers.back()));
+      values_.insert(values_.end(), page.begin(), page.end());
+    }
+  }
+
+  void VerifyOutput(size_t total_values) {
+    TypedScanner<TYPE>* scanner = reinterpret_cast<TypedScanner<TYPE>* >(scanner_.get());
+    T val;
+    bool is_null;
+    size_t j = 0;
+    for (size_t i = 0; i < total_values; i++) {
+      ASSERT_TRUE(scanner->HasNext());
+      ASSERT_TRUE(scanner->NextValue(&val, &is_null));
```

Comment:
The `Scanner` API is deficient as you cannot scan the values and levels (as a user) at the same time if I'm reading the code right. Can you open a JIRA to add a `Scanner::Next(int16_t* rep_level, int16_t* def_level, T* val, bool* is_null)` API? We can return and improve the test case as part of that JIRA